explorer/src/providers/anchor.tsx

```diff
@@ -0,0 +1,49 @@
+import { Idl, Program, Provider } from "@project-serum/anchor";
+import { Connection, Keypair } from "@solana/web3.js";
+import { NodeWallet } from "@metaplex/js";
+
+const cachedAnchorProgramPromises: Record<
+  string,
+  | void
+  | { __type: "promise"; promise: Promise<Program<Idl> | null> }
+  | { __type: "result"; result: Program<Idl> | null }
+> = {};
+
+export function useAnchorProgram(
+  programAddress: string,
+  url: string
+): Program | null {
+  const key = `${programAddress}-${url}`;
+  const cacheEntry = cachedAnchorProgramPromises[key];
+
+  if (cacheEntry === undefined) {
+    const promise = Program.at(
+      programAddress,
+      new Provider(new Connection(url), new NodeWallet(Keypair.generate()), {})
+    )
+      .then((program) => {
+        cachedAnchorProgramPromises[key] = {
+          __type: "result",
+          result: program,
+        };
+        return program;
```

Comment:
No need to return anything; the `Promise` is just a signal to React to retry the render.

Reply:
addressed in latest commit, thanks for catching!